### ISSUE TITLE
Evaluation

### ISSUE THREAD

**Issue by vidbap** [NisaarAgharia/Indian-LawyerGPT#4]
Hello,
First of all great work and thanks for sharing - you are democratizing AI to newbies.  I tried with my own dataset formatted like Q-A pairs and ran the generated model to infer back on training data itself to find out how much it learn't.  I didn't get good results in the sense that it used general knowledge to answer and not the QA I trained on.  Any advice?

How is your evaluation on running the inference on training data?  It would be interesting to know if there is a metrics which captures how much a model learn't from training data.  Thanks,

Viba

**Comment by ajoyzz** [comment]
More updates.  With temperature = .1 at inference, it gave better results than using the default .7 in the code.  This can improve answer quality but still lacking the TRUTH, ie. I get more creative answers even if I give the same question that was used in training set.  Any insights greatly appreciated.

Another surprise is that we trained a few other models and are finding similar anomalies in the output where the model is not truthful to the data it is trained on.

**Comment by NisaarAgharia** [comment]
The temperature 0 will make the model most deterministic and will not create random output. 

Ideally it should create something on the lines of trained data.

What are the hyper params you used for training the model
for eg

Number of training examples? 
Number of epochs?
Max steps?

**Comment by ajoyzz** [comment]
Thanks Nissar for your guidance.  Yes, with temperature = 0, we got the style and a few phrases repeated from training data but yet, not a faithful reproduction of training data.  It seems to merge information from pretrained model and information trained for when it generates.

Summary:  (A100)
Training data = 682 rows of data (1364 turns)
Epochs = 5
Batch_size=2
le rate = 2e-4

Do you think these parameters should be changed to improve model accuracy?

- Increase epochs (but danger of overfitting?). May be overfitting is good for this use case.
- Lower le/Higher le

I used pretty much the parameters you had in the original code in this repo.  See below for details.

Training data had 682 pairs of questions and answers
Dataset({
    features: ['question', 'answer'],
    num_rows: 682
}) which gets converted to this: (682*2 turns)
<human>: How do you configure password authentication for the review phase?
<assistance>:  You can configure password authentication to authenticate reviewers on the approval banner in Smart UI. You can set password authentication as per review phases, and if required, you can skip password authentication for a particular review...
...
training_args = transformers.TrainingArguments(
    per_device_train_batch_size=2,
    gradient_accumulation_steps=4,
    num_train_epochs=5,
    learning_rate=2e-4,
    fp16=True,
    save_total_limit=4,
    logging_steps=10,
    output_dir=OUTPUT_DIR,
    max_steps=800,
    optim="paged_adamw_8bit",
    lr_scheduler_type = 'cosine',
    warmup_ratio = 0.05,
    report_to = 'tensorboard'
)
My loss was nicely decreasing;

770 | 0.979100
780 | 0.875500
790 | 0.991000
800 | 0.984100
TrainOutput(global_step=800, training_loss=1.717551646232605, metrics={'train_runtime': 3279.0019, 'train_samples_per_second': 1.952, 'train_steps_per_second': 0.244, 'total_flos': 5.185419995011277e+16, 'train_loss': 1.717551646232605, 'epoch': 9.38})

BnB & Lora config also same as yours:
MODEL_NAME = "tiiuae/falcon-7b-instruct"

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    load_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16,
)

model =AutoModelForCausalLM.from_pretrained(
    MODEL_NAME,
    device_map="auto",
    trust_remote_code=True,
    quantization_config=bnb_config,
)
config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["query_key_value"],
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM"
)

**Comment by NisaarAgharia** [comment]
Also we need loss to go in the range of 0.2 to 0.05 currently your loss at 800 steps is around 0.9 which is high

**Comment by vidbap** [comment]
You are brilliant.  After I fine tuned with 3600 steps, I got less than .05 loss.  The test dataset is also now giving a **lot better answers** compared to ground truth - about 40% of my original FAQ questions are answered truthfully, which is quite good. 

- The SOTA for Natural Questions is 40-60 accuracy as per https://paperswithcode.com/sota/question-answering-on-natural-questions  
- https://huggingface.co/spaces/HuggingFaceH4/open_llm_leaderboard best models are 50- range
- https://crfm.stanford.edu/helm/latest/?group=natural_qa_closedbook also has highest scoring models at .4 F1

1. Should we conclude that the best we can get is 40-60% accuracy in terms of factual accuracy?  How do we further improve?  Any learning rate changes like you suggested #steps?
2. How do we ensure that the model has not overfit?  Should we add evaluation loss to check both training and eval dataset loss curves?  But not sure if we can do this for Causal LM.
3. Finally, after this model is trained, is there a way we can do a “SEQ_2_SEQ_LM” task_type to further fine tune with a sequence to sequence LM where we are making it TASK specific to predict answers from questions.

Just throwing out some ideas on what you think and which option has best success path forward.  Once again, appreciate your answers in this learning journey.  Best.

**Comment by meghantra** [comment]
It’s great to see you experimenting with your own dataset—using Q-A pairs is an excellent way to get started with fine-tuning models.

Regarding your observation about the model relying on general knowledge rather than the specific Q-A pairs you trained it on, this could happen for a few reasons:

Insufficient Training Data: If the dataset is too small or not diverse enough, the model might not fully capture the nuances of your Q-A pairs and fall back on pre-trained general knowledge. Adding more high-quality examples that align with your use case could help.

Learning Rate and Epochs: Fine-tuning requires balancing between underfitting and overfitting. If the learning rate is too high or the number of epochs is too low, the model might not learn your specific data effectively.

Model Architecture: Some pre-trained models are optimized for general-purpose tasks. If you’re fine-tuning for a highly specific task (e.g., legal queries), consider using a smaller, domain-specific model.

When evaluating how much a model has learned, running inference on the training data is one way, but it’s not the best metric. It often overestimates performance because the model has already seen this data. Instead, you should consider:

Validation Data: Split your dataset into training and validation sets and evaluate the model on unseen validation data to get a realistic sense of its learning.
Metrics: Use metrics like F1-score, BLEU (for language models), or exact match accuracy to evaluate how well the model aligns with your Q-A pairs.
On a side note, I’ve worked on projects that required highly specific legal expertise similar to how one might consult a[ Houston appeal attorney](https://www.brownstonelaw.com/appellate-lawyer/texas-appeals/houston-texas-appellate-law-firm/
) for handling complex appellate cases. In AI, having a specialized dataset is like having that expert attorney: it ensures precision and relevance. Fine-tuning your model with domain-specific data and validating it rigorously can help you achieve better results.
